⚡Dataset + DataLoader for sampling

Convolutio opened the pull request:
✨ IterableDataset for sampling ready to be used in multithreading context.

Screamnox committed:
♻️📝 `extractIPA.py`

Screamnox committed:
♻️📝`data` folder
(+ adapted outside files)

Screamnox committed:
🎨aaaaaaaa
(just some a)

Screamnox committed:
🐛Fix `padDataToNgram` and `train`
- Switch in vocab parenthesis indices
- Correct conditions for mask
- Changed `smoothingValue` (log) to `1e-5` instead of `-1e5`
- Invert where elements for positionToSmooth

Screamnox committed:
🐛Revert `smoothingValue`

Screamnox committed:
🐛Revert mask
(conditions was good)

Convolutio committed:
first_dataset

Convolutio committed:
🚧new dataset

Convolutio committed:
🧪more realistic simulation
realistic samples length
torch.uint8 dtype for memory economy (3 Go)

Convolutio committed:
🐛renamed imports + sampling notebook

Screamnox committed:
🐛Correct sum by logsumexp
(+ Update prior_lm.ipynb)

Screamnox committed:
💩Fix name folders + tensor device
(Fast fix, need to do correctly)

Convolutio committed:
Merge branch 'dev' into sampling_dataset
update the state of Article_Scientifique and Mémoire folders

Convolutio committed:
⚡IterableDataset available for multithreading
validated tests on notebooks for the mini-batches
distribution over each worker.

Convolutio committed:
🗒️import naming  correction

Screamnox commented on Code/Unsupervised_reconstruction/Source/dynamicPrograms.py:
```diff
@@ -3,13 +3,13 @@
 BIG_NEG = -1e9
 
-
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
```
It was added quickly to carry out sampling tests, but deserves further thought, in particular to ensure that vectors are created on the right device.

TDLR; Currently ok, but in the future should be removed and the initialisation of vectors on the chosen device should be standardised.

Screamnox replied:
It was added quickly to carry out sampling tests, but deserves further thought, in particular to ensure that vectors are created on the right device.

Currently ok, but in the future should be removed and the initialisation of vectors on the chosen device should be standardised.

Screamnox commented on Code/Unsupervised_reconstruction/Source/dynamicPrograms.py:
```diff
@@ -30,8 +30,8 @@ def compute_mutation_prob(model:EditModel, sources_:InferenceData_SamplesEmbeddi
-    f_sub = torch.full((max_source_sequenceLength, max_target_sequenceLength, *batch_shape), fill_value=BIG_NEG)
-    f_ins = torch.full((max_source_sequenceLength, max_target_sequenceLength, *batch_shape), fill_value=BIG_NEG)
+    f_sub = torch.full((max_source_sequenceLength, max_target_sequenceLength, *batch_shape), fill_value=BIG_NEG, device=device)
+    f_ins = torch.full((max_source_sequenceLength, max_target_sequenceLength, *batch_shape), fill_value=BIG_NEG, device=device)
```
When the global variable device disappears, these arguments will no longer be valid.

Screamnox commented on Code/Unsupervised_reconstruction/Source/editModel.py:
```diff
@@ -142,7 +142,7 @@ def convertForIndexing(t): return torch.cat(
         x_l, y_l, c, b = sub_results.shape[:-1]  # |x|+1, |y|+1, c, b
         # neutralizes results for the padding and eos tokens
-        not_padding_token_in_source = (torch.arange(sources[2]-1)[:, None, None] < (sources[1]-1).unsqueeze(0)).unsqueeze(1).unsqueeze(-1) # shape = (|x|+1, 1, c, b, 1)
+        not_padding_token_in_source = (torch.arange(sources[2]-1)[:, None, None] < (sources[1]-1).unsqueeze(0)).unsqueeze(1).unsqueeze(-1).to(device) # shape = (|x|+1, 1, c, b, 1)
```
This transfer to the desired `device` has an impact on performance, and should not be carried out here. 
In addition, the global `device` variable will disappear in future.

Screamnox commented on Code/Unsupervised_reconstruction/Source/editModel.py:
```diff
@@ -172,12 +172,11 @@ def cache_probs(self, sources: InferenceData_SamplesEmbeddings, targets: Inferen
-                torch.zeros((t.size()[0]+1, padding_y,
-                            *t.size()[2:]), device=device)
-            ), dim=1)
+            def lengthen(t, padding_x, padding_y): 
```
Review this function, which seems to be costly in terms of performance due to the number of times it is called, in particular by device transfer.

Screamnox commented on Code/Unsupervised_reconstruction/Source/generateProposals.py:
```diff
@@ -18,7 +18,7 @@
 
 
-def computeMinEditDistanceMatrix(x: Tensor, y: Tensor) -> npt.NDArray[np.int_]:
+def __computeMinEditDistanceMatrix(x: Tensor, y: Tensor) -> npt.NDArray[np.int_]:
```
Shouldn't the PyTorch be the standard?

_We're only talking about standards here, not performance, even though it may perform better in the current library._

Screamnox commented on Code/Unsupervised_reconstruction/Source/generateProposals.py:
```diff
@@ -166,19 +166,19 @@ def sampleFromProposals(proposalsSet:Tensor, samplesNumber:int)->Tensor:
-        Y = [cognates[language][i] for language in cognates]
-        p.append(sampleFromProposals(computeProposals(x, Y), samplesNumber))
+        Y = list(cognates[i].values())
+        p.append(__sampleFromProposals(__computeProposals(x, Y), samplesNumber))
```
We can dispense with the conversion to `list` of the values of each of the `cognates[i]` dictionary. 

Indeed, it does not correspond to the type expected by the `__computeProposals` function, which expects to receive a list of tensors for the `cognates` argument, but we can see that the function only uses the argument for a 'for in' loop, which implicitly converts the list back into an iterator. 

So, for performance and memory reasons (however small they may seem), it would be better to remove. 

```python
def __computeProposals(currentReconstruction: Tensor, cognates: list[Tensor]) -> Tensor:
    ...
    for cognate in cognates:
        ...
```

Screamnox commented on Code/Unsupervised_reconstruction/Source/reconstructionModel.py:
```diff
@@ -9,14 +9,14 @@
 import matplotlib.pyplot as plt
 
-from models.types import (ModernLanguages, Operations, InferenceData,
+from models.types import (ModernLanguages, Operations, InferenceData_Samples,
```
Shouldn't we create classes instead of aliases for these objects?

For instance, currently, when trying to access elements of these objects (typed by the alias `InferenceData_Samples`), the operation is performed using `object[0]`, `object[1]`, etc., as it is treated like a list. However, through a class (which can inherit from a list to retain current properties), we could manipulate these objects more clearly by implementing attributes with names.

The aim is to enhance readability and ease the manipulation of objects typed with this class. Nevertheless, it is essential to assess the memory impact of this implementation, even though I am confident it will be negligible.

Screamnox commented on Code/Unsupervised_reconstruction/Source/reconstructionModel.py:
```diff
@@ -9,14 +9,14 @@
+from Source.dynamicPrograms import compute_mutation_prob, compute_posteriors
 from models.probcache import ProbCache
 
 device = 'cuda' if torch.cuda.is_available() else 'cpu'
```
In connection with my previous comments on devices, the one and only `device` declared as a global variable should be located here. (Excluding the _LMs_).

Indeed, all calls leading to various files in the project (source folder) should rely on the device imposed/passed by the `ReconstructionModel`. 
And even remove this global variable to be passed as an argument during the model initialization.

Even though, in this development phase, not all files are yet perfect, the ultimate goal should lean towards this. Can you confirm @Convolutio, or are there files that are highly independent of `ReconstructModel`?

Convolutio replied:
For the one that I remember, there are:
- Source/dynamicPrograms.py
- Source/editModel.py
- data/vocab.py (all the `ComputeInferenceData_*` method returns the input for the reconstruction model and then the device should be defined there)
- as a consequence of the input tensors' device computed in the file above, the prior language models should also have a device related to `ReconstructionModel`. A solution may be to incorporate the prior language model in the `ReconstructionModel` instances so that the device can be internally set up.

Screamnox commented on Code/Unsupervised_reconstruction/Tests/createSamples.py:
```diff
@@ -1,23 +1,21 @@
-    
-torch.save(batch, './tests/samples.pt')
\ No newline at end of file
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
```
Note: To support my perspective on devices, it is justified to keep the variable here because it is _independent_ of the model; it does not originate from the model. And I am not referring to checking the same device between the model and pre-model data.

Screamnox commented on Code/Unsupervised_reconstruction/Tests/timeToSample.py:
```diff
@@ -1,4 +1,4 @@
-from source.generateProposals import computeProposals, computeMinEditDistanceMatrix
+from Source.generateProposals import computeProposals, computeMinEditDistanceMatrix
```
- `computeProposals` -> `__computeProposals` ?
- `computeMinEditDistanceMatrix` -> `__computeMinEditDistanceMatrix` ?

Screamnox replied:
Private methods, so no longer accessible.

Screamnox commented on Code/Unsupervised_reconstruction/data/reconstruction_datasets.py:
```diff
@@ -23,7 +25,10 @@ def __training__collate_fn(batch:list[tuple[str, dict[ModernLanguages, str], __C
     return (firstElement, secondElement, lastElement)
 
-class __TrainingDataset(Dataset):
+class __TrainingDataset(Dataset[tuple[str,
```
Is it not too lengthy for a type? Would a local alias before the definition not be justified here?

Convolutio replied:
the next commit has fixed that 👍

Screamnox commented on Code/Unsupervised_reconstruction/data/reconstruction_datasets.py:
```diff
@@ -34,7 +39,7 @@ def __init__(self, raw_samples: list[str],
         return self.length
     
-    def __getitem__(self, index):
+    def __getitem__(self, index: int):
```
Forget about the return type.

Screamnox replied:
Fixed! 👍

Screamnox commented on Code/Unsupervised_reconstruction/data/reconstruction_datasets.py:
```diff
@@ -48,4 +53,57 @@ def trainingDataLoader(raw_samples:list[str],
 
 
 #-------------Sampling Datasets---------------------
+Raw_minibatch_type = tuple[tuple[
```
Exactly!
(In relation to the act of using an alias before the class.)

Screamnox commented on Code/Unsupervised_reconstruction/data/reconstruction_datasets.py:
```diff
@@ -48,4 +53,57 @@ def trainingDataLoader(raw_samples:list[str],
+            num_workers = worker_info.num_workers
+        samples_batch_index = worker_id
+        for start_index in range(worker_id*self.batch_size, len(self.proposals_dataset), num_workers*self.batch_size):
+            yield((self.proposals_dataset[start_index:start_index+self.batch_size], self.cognates), (self.cognates_batch_index, samples_batch_index))
```
Appears to contain an error in the yield.
(Need more time to be certain.)

Screamnox replied:
After re-reading, the expression for _yield_ seems OK. However, the type returned by `self.proposals_dataset[start_index:start_index+self.batch_size]` is a `tuple[Tensor, ...]` and not a `list[Tensor]`.

@Convolutio, can you elaborate on your perspective regarding the use of this expression?

Convolutio replied:
The raw proposal mini-batch got from the `TensorDataset` instance with this instruction :
> `self.proposals_dataset[start_index:start_index+self.batch_size]`

is in the format of a tuple of $c$ tensors of shape $(b, |x|)$, with $b$=`self.batch_size`. $c$ is the number of cognate pairs given in the `__SamplingDataset` instance. Afterwards, this tuple is being able to be collated in the `__sampling__collate_fn` function.

`start_index` has been set up so that two distinct workers do not load any common mini-batch.

Screamnox commented on Code/Unsupervised_reconstruction/data/reconstruction_datasets.py:
```diff
@@ -34,7 +39,7 @@ def __init__(self, raw_samples: list[str],
+    def __getitem__(self, index: int):
         return self.training_load[index]
 
 def trainingDataLoader(raw_samples:list[str],
```
The return seems inconsistent between its definition and the current object.

Screamnox commented on Code/Unsupervised_reconstruction/data/reconstruction_datasets.py:
```diff
@@ -48,4 +53,57 @@ def trainingDataLoader(raw_samples:list[str],
+                        tuple[int, int] # the coords of the minibatch in the global batch
+                        ]
 
+class __SamplingDataset(IterableDataset[Raw_minibatch_type]):
```
Why the choice of an `IterableDataset`?

Convolutio replied:
The global dataset does not have to be entirely loaded to transform and process the sequence of mini-batches. Moreover, the cutting in mini-batches does not require any shuffling operation and can be carried out in a sequential order. The `IterableDataset` class enables to control the process flow in this manner with letting a huge dataset be loaded within, for instance by chaining several instances of it (see the usage of `ChainingDataset` in this script for an example).

I concede the better usage would be to just give the filename of a .pth for constructing a `__SamplingDataset` instance. If we note C the total number of cognate pairs and B the number of proposals to be processed per cognate pair, this file would contain here a sublist of c tensors with B proposals. Then, if the C tensors of B proposals had had to be splitted into sublists of c tensors of B proposals for not saturing the memory, a list of ~C/c files --- each containing a sublist --- would be able to be loaded in a chained dataset of `__SamplingDataset`s.  
But in our case we do not need to load several files and can use the whole generated list of C tensors in memory, so the IterableDataset is only used for loading the mini_batches approximately in the order that fits their coordinates (if several workers are deployed, multiprocessing may break a bit this order but it does not matter at all).

Screamnox replied:
IterableDatasets are perfect for this situation. And in our case, as you mentioned, there's no need to create files for each instance of __SamplingDataset, they can be stored directly in memory.

Screamnox commented on Code/Unsupervised_reconstruction/lm/PriorLM.py:
```diff
@@ -213,7 +213,7 @@ def training(self, data: str, mini_batch_size: int = 32, epochs: int = 10, save_
         adjust_seq_lengths: Callable[[Tensor, Tensor, int], tuple[Tensor, Tensor]] = lambda x, l, _: (x, l+1)
         training_data = [adjust_seq_lengths(
-            *computeInferenceData(tData)) for tData in indices_tensor.split(mini_batch_size, dim=1)]
+            *computeInferenceData_Samples(tData)) for tData in indices_tensor.split(mini_batch_size, dim=1)]
```
Wouldn't there be an operation to speed up the procedure? Especially to avoid the need for `adjust_seq_lengths` through a Torch method?

Convolutio replied:
The module is not updated with the last changes (for instance in the `inference` method and a review will have to be done).

Screamnox reviewed (commented):
Can carry out the merge, subject to the following:
- Some aspects need to be reconsidered to enhance performance and memory, even if the improvements may seem minor. 
- At first glance, the new section on DataLoader & Dataset for sampling appears to be correct; however, it requires a review regarding the types of data being processed.

Screamnox commented on Code/Unsupervised_reconstruction/data/reconstruction_datasets.py:
```diff
@@ -1,8 +1,10 @@
+from data.vocab import computeInferenceData_Samples, computeInferenceData_Cognates, wordsToOneHots, vocabulary
+from Source.utils import pad2d_sequence
+from models.types import ModernLanguages, Operations, InferenceData_Samples, InferenceData_Cognates, PADDING_TOKEN
 
```
After a more thorough reading of the _"reconstruction_datasets.py"_ file, it is necessary to review the data flow and structure, in particular to avoid :
- the costs of alternating between object list tuples and object tuple lists
- duplication of data, as found in the file which sends duplicate `dict[ModernLangues, Any]`.

In addition, this file shows the need to implement classes for these objects to make them easier to handle and read.

Screamnox reviewed (commented):
After a more thorough reading of the _"reconstruction_datasets.py"_ file, it is necessary to rework the data flow and structure, in particular to avoid :
- the costs of alternating between object list tuples and object tuple lists
- duplication of data, as found in the file which sends duplicate `dict[ModernLangues, Any]`.

In addition, this file shows the need to implement classes for these objects to make them easier to handle and read.

Screamnox committed:
♻️New aliases + Correct types
(TODO: Replace these aliases by classes!)

Screamnox committed:
♻️Revert name folders

Screamnox committed:
🚧Merge branch 'sampling_dataset' into 'restruct'
(TODO: Correct type errors, notebooks, and test files)

Screamnox committed:
🎨

Screamnox committed:
🐛Revert sum
(do not produced the expected result)

Convolutio committed:
✏️renamed types + 🐛fixed length typo

Screamnox committed:
🚧Update PriorLM notebook
(add new stats + graphs)

Screamnox committed:
✨`checkSameDevice`

Screamnox committed:
🚧♻️Unifomise device
- Not tested!
- Refactor to ensure the uniformisation of all device
(- In addition to a few rewrites)

Screamnox committed:
🏗️♻️Uniformise device
(this commit can be ignored when merging with `sampling_dataset`)

Convolutio committed:
🐛 shape error

Screamnox committed:
💩Empty commit
(test changed)

Screamnox committed:
Revert "💩Empty commit"

This reverts commit 89fe1823437ac1178219e1bbcf1970edf30fc7b3.

Convolutio committed:
🐛 type error + edited some notebooks

Convolutio committed:
Merge remote-tracking branch 'origin/restruct' into sampling_dataset

Convolutio committed:
🐛 fixed devices issue (@Screamnox)
